sql: acquire change lease during CREATE..AS backfill

dt opened the pull request:
Only one instance of the CREATE TABLE ... AS backfill should only be run at a time since the resulting
table should contain the output of exactly one run of the souce query (i.e. since it could produce
difference results, the output doesn't have a PK to de-duplicate, etc). Previously, based on where it
was being called in the schema change exec() method however, it was executing without holding the
schema-change lease, meaning other nodes that saw the schema change could (and did) attempt to execute
it as well, at the same time, leading to incorrect results.

It is possible that the entirety of schemaChanger.exec should be executed while holding the lease to
prevent similar mistakes in the future, however currently some methods acquire the lease themselves
while others rely on it being acquired for them.

Based on where the backfill happens, it is expected to acquire the lease itself if it needs it, so this
change does just that. Additionally, it runs the backfill in one goroutine while running another to
maintain said lease, ensuring that it should remain held even if the backfill takes longer than the
lease expiration time.

Release note (bug fix): fix a bug where multiple nodes attempted to populate the results for CREATE TABLE ... AS leading to duplicate rows."

dt committed:
sql: add logging of when schema changes start executing

Schema changes are relatively uncommon events, and are often significant ones, where issue or bugs are more likely.
Thus, it logging during a schema change is likely to be helpful when trying to diagnose an issue, and is unlikely
to be too noisy given their infrequent nature.

This change adds logging around when a schema change begins exeution, as many things -- like ordering of changes or leases --
affect if/when it runs, and it is helpful to be able to determine if and when it actually did so.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43840)
<!-- Reviewable:end -->

dt committed:
sql: acquire change lease during CREATE..AS backfill

Only one instance of the CREATE TABLE ... AS backfill should only be run at a time since the resulting
table should contain the output of exactly one run of the souce query (i.e. since it could produce
difference results, the output doesn't have a PK to de-duplicate, etc). Previously, based on where it
was being called in the schema change exec() method however, it was executing without holding the
schema-change lease, meaning other nodes that saw the schema change could (and did) attempt to execute
it as well, at the same time, leading to incorrect results.

It is possible that the entirety of schemaChanger.exec should be executed while holding the lease to
prevent similar mistakes in the future, however currently some methods acquire the lease themselves
while others rely on it being acquired for them.

Based on where the backfill happens, it is expected to acquire the lease itself if it needs it, so this
change does just that. Additionally, it runs the backfill in one goroutine while running another to
maintain said lease, ensuring that it should remain held even if the backfill takes longer than the
lease expiration time.

Release note (bug fix): fix a bug where multiple nodes attempted to populate the results for CREATE TABLE ... AS leading to duplicate rows.

thoszhang reviewed (approved):
> Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43840)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/sql/schema_changer.go, line 1226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43840#-LyWHPDP0ZMlQzDwac-j:-LyWHPDP0ZMlQzDwac-k:bl2i0s9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7820f3580965b3fc1c358a9534b65a1d91b6928e/pkg/sql/schema_changer.go#L1226)):*
> ```Go
> 	_, err := sc.leaseMgr.WaitForOneVersion(ctx, tableID, retryOpts)
> 	if log.V(2) {
> 		log.Infof(ctx, "waiting for a single version of table %d... done", tableID)
> ```

Should we also unconditionally log this? (We could even get the `DescriptorVersion` from `WaitForOneVersion` instead of ignoring it, and add that to the log statement.)


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
TRTRs

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43840#-:-LzFZmIgBylWGjmBL9yB:b61kby1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/schema_changer.go, line 1226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43840#-LyWHPDP0ZMlQzDwac-j:-LzFZVXIBSRR5lJ8KODl:bpz1bqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7820f3580965b3fc1c358a9534b65a1d91b6928e/pkg/sql/schema_changer.go#L1226)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Should we also unconditionally log this? (We could even get the `DescriptorVersion` from `WaitForOneVersion` instead of ignoring it, and add that to the log statement.)
</blockquote></details>

:thumbsup: 

Done.


<!-- Sent from Reviewable.io -->
